projects/telemetry/src/gps.c

```diff
@@ -0,0 +1,194 @@
+#include "gps.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include "gpio.h"
+#include "interrupt.h"
+#include "log.h"
+#include "misc.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "uart.h"
+#include "uart_mcu.h"
+#include "xbee.h"
+
+// Just some constants so that the max length of raw data can be set.
+// A GGA message will be around a hundred characters.
+#define s_max_nmea_length 128
+
+static UARTSettings s_gps_settings = {
+  .baudrate = 9600,
+  .tx = { .port = GPIO_PORT_A, .pin = 2 },
+  .rx = { .port = GPIO_PORT_A, .pin = 3 },
+  .alt_fn = GPIO_ALTFN_1,
+};
+
+// The pin numbers to use for UART
+static const GPIOAddress s_pins[] = {
+  { .port = GPIO_PORT_B, .pin = 3 },  // Pin GPS power
+  { .port = GPIO_PORT_B, .pin = 4 },  // Pin GPS on_off
+};
+
+// The UART port to use for the GPS
+static const UARTPort s_gps_port = UART_PORT_2;
+
+static GPIOSettings s_settings_gpio_general = {
+  .direction = GPIO_DIR_OUT,  // The pin needs to output.
+  .state = GPIO_STATE_LOW,    // Start in the "off" state.
+  .alt_function = GPIO_ALTFN_NONE,
+};
+
+// Keeps track of whether the GPS is sending data or not
+static volatile bool s_gps_active = false;
+
+// Keeps track of whether we want the GPS to be active or not
+static volatile bool s_gps_desired_state = true;
+
+// Stores raw NMEA messages sent by the chip
+static volatile char s_gga_data[s_max_nmea_length];
+
+static gps_settings s_settings = { .settings_power = &s_settings_gpio_general,
+                                   .settings_on_off = &s_settings_gpio_general,
+                                   .pin_power = &s_pins[0],
+                                   .pin_on_off = &s_pins[1],
+                                   .uart_settings = &s_gps_settings };
+
+static UARTStorage s_storage;
+
+static bool s_gps_nmea_is_gga(const char *to_check) {
+  if (to_check == NULL) {
+    return false;
+  }
+  size_t len = strlen(to_check);
+  if (len < 6) {
+    return false;
+  }
+  if (to_check[3] == 'G' && to_check[4] == 'G' && to_check[5] == 'A') {
+    return true;
+  }
+  return false;
+}
+
+// This method will be called every time the GPS sends data.
+static void prv_gps_callback(const uint8_t *rx_arr, size_t len, void *context) {
+  if (s_gps_nmea_is_gga((char *)rx_arr)) {
+    snprintf(s_gga_data, s_max_nmea_length, "%s", (char *)rx_arr);
+    s_gps_active = true;
+  }
+}
+
+// The GPS power line will be connected to a 3V pin. This method will
+// control if that pin is providing power or not.
+static void prv_gps_set_power_state(gps_settings *settings, bool powered) {
```

Comment:
`settings` should be a `const` pointer and your type should be `GpsSettings`....